### DISCUSSION TITLE
G-Code Layer number

### DISCUSSION THREAD

**Discussion by Denniswie** [supermerill/SuperSlicer#1797] [Q&A] [answered]
When looking at the G-code preview, the base layer is '1'
When looking at the G-code file, the first Layer is '0' ( M117 Layer 0 ........) ( reference "  ; before_layer_gcode = M117 Layer [layer_num], z = [layer_z]\n   "
- is it possible to amend SuperSlicer so the G-code generated starts at '1' ? eg:  ( M117 Layer [layer_num + 1], z =......)
or
- is there a post process to go through the generated code, and add 1 to the layer counts. ?

Its causing me a little confusion, when I insert my colour change script.
Also, my OCD is offended :-)

**Comment by supermerill** [comment]
Everything in software computing began at 0.
It's just for the GUI that it began at 1, to avoid confusing normal people.
In gcode, index has to start at 0, or I'll lose 99% of the ones that use them.

in your custom gcode, you can do math, so you can "+1" things.

**Reply by chuan2984** [reply]
not in lua :)

**Comment by Denniswie** [comment]
I've been busy in the layer change G-code , and google G-Code script language, but struggling, can you suggest good reference for language.

I've so far tried all combinations  ( {  [
M117 Layer {[layer_num] + 1}, z =.....
M117 Layer [layer_num] +1, z =.....
etc.
All the places I visited dont seem to do maths with the parameters

**Reply by supermerill** [reply]
https://www.help.prusa3d.com/en/article/macros_1775

`{layer_num + 1}`

**Comment by Denniswie** [comment, marked as answer]
OK, got it.
M117 Layer {layer_num +1}, z = [layer_z]

The one combination I did not think of - stripping off the [ ] square brackets
Many thanks for link. What a goo resource.

**Reply by supermerill** [reply]
`[]` are not useful, they are still available to be able to compute old macro, but [truc] is just a shortcut for {truc} or {truc[currrent_extruder]} if it's an array.

**Comment by Denniswie** [comment]
I need to know if there is a placeholder for the total number of layers.
I have checked the Slic3r manual, (  https://manual.slic3r.org/advanced/placeholder-parser  ) and there seems to be only
 "layer_num" - current active layer.
I would like to modify on layer change g-code to be:
M117 Layer [layer_num] of [total_layers] layers
So I guess my question is - Is the list of placeholders in the manual all there is ?

**Reply by supermerill** [reply]
https://github.com/supermerill/SuperSlicer/wiki/Variable-list
`total_layer_count`
